plugins/shadow-block-converter/src/shadow_block_converter.js

```diff
@@ -0,0 +1,194 @@
+  }
+
+  /**
+   * Run a change event.
+   * @param {boolean} forward True if run forward, false if run backward (undo).
+   * @override
+   */
+  run(forward) {
+    const workspace = this.getEventWorkspace_();
+    const block = workspace.getBlockById(this.blockId);
+    if (!block) {
+      console.warn('Can\'t change non-existent block: ' + this.blockId);
+      return;
+    }
+
+    const value = forward ? this.newValue : this.oldValue;
+    block.setShadow(!!value);
+  }
+}
+
+Blockly.registry.register(
+    Blockly.registry.Type.EVENT,
+    BlockShadowChange.EVENT_TYPE,
+    BlockShadowChange);
+
+/**
+ * Add this function to your workspace as a change listener to automatically
+ * convert shadow blocks to real blocks whenever the user edits a field on the
+ * block, like this:
+ *
+ * workspace.addChangeListener(shadowBlockConversionChangeListener);
+ *
+ * Ideally the Blockly.Field.prototype.setValue method should handle this logic,
+ * but for the purposes of this plugin it can be a workspace change listener.
+ *
+ * @param {Blockly.Events.Abstract} event An event broadcast by the workspace.
+ */
+export function shadowBlockConversionChangeListener(event) {
+  // Auto-converting shadow blocks to real blocks should happen in response to
+  // new user action events (which get recorded as undo events) but not when
+  // undoing or redoing events (which do not get recorded again).
+  if (!event.recordUndo) {
+    return;
+  }
+
+  // Auto-converting shadow blocks to real blocks should happen in response to
+  // editing a field value, which is recorded as Blockly.Events.BLOCK_CHANGE.
+  if (event.type != Blockly.Events.BLOCK_CHANGE) {
+    return;
+  }
+
+  const workspace = Blockly.Workspace.getById(event.workspaceId);
+  if (!workspace) {
+    return;
+  }
+  const block = workspace.getBlockById(event.blockId);
+  if (!block) {
+    return;
+  }
+
+  // Blocks that are already real blocks can be ignored.
+  if (!block.isShadow()) {
+    return;
+  }
+
+  // Remember the current event group so that it can be resumed below.
+  const currentGroup = Blockly.Events.getGroup();
+
+  if (event.group) {
+    // Temporarily use the same group as the initiating event so that
+    // the shadow events get grouped with it for undo purposes.
+    Blockly.Events.setGroup(event.group);
+  } else {
+    // The initiating event wasn't part of any named group, so the shadow events
+    // can't be grouped with it, but at least they can be grouped with each
+    // other. This happens when editing e.g. boolean fields.
+    Blockly.Events.setGroup(true);
+
+    // HACK: Retroactively overwrite the initiating event's group property in
+    // order to include it in the same group as the shadow events anyway. It
+    // wasn't previously grouped with any other events, but now it can be
+    // grouped with the new events.
+    event.group = Blockly.Events.getGroup();
+  }
+
+  // If the changed shadow block is, itself, a child of another shadow block,
```

Comment:
What if the block is a shadow and has *children* that are shadows? For instance, if I change the dropdown on the first addition block here:
![image](https://user-images.githubusercontent.com/13686399/192910822-59ea696c-8ab6-4c1e-8478-228af9134b37.png)

Reply:
I chose to only ensure that all parents of the changed block are real, and ignore the children. As you pointed out, "Doing it this way still maintains the invariants that no shadow block has a real child, while changing the minmum number of blocks."

Actually, that invariant is temporarily violated while in the middle of creating shadow change events, because the algorithm starts converting shadow blocks at the one that's edited before walking up the hierarchy. I guess technically it should walk up the hierarchy first, then start reifying blocks starting at the top and working back down to the one that was edited. 

In my testing it seems to work just fine so far. If the events are always executed as a group, then I assume nobody will notice. But, hypothetically, if the blockly source code is changed to depend on maintaining this invariant at all times then this plugin could get broken.

Reply:
> Is it safe to assume that every block will only have one parent (an outputConnection or a previousConnection but not both)? 

Yes, you can assume this. Every now and then someone asks to change it, but we've held the line :)

Reply:
Great, thanks! I've committed a version that reifies the parent blocks first, along with a unit test that enforces the order of shadow block events.

Reply:
> Yes, you can assume this. Every now and then someone asks to change it, but we've held the line :)

Just to clarify, it is possible to have both an output and a previous connection! But only one of them can be connected to a parent at any given time.